kamu-adapter-http/src/http_server_simple_transfer_protocol.rs

```diff
@@ -0,0 +1,229 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::{
+    ws_smart_transfer_protocol_axum_server, DatasetResolver, PARAMETER_BLOCK_HASH,
+    PARAMETER_DATASET_NAME, PARAMETER_PHYSICAL_HASH, PARAMETER_REF,
+};
+use axum::extract::Extension;
+use kamu::domain::*;
+
+use opendatafabric::{
+    serde::{flatbuffers::FlatbuffersMetadataBlockSerializer, MetadataBlockSerializer},
+    Multihash,
+};
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+use url::Url;
+
+/////////////////////////////////////////////////////////////////////////////////
+
+pub async fn dataset_refs_handler(
+    dataset: Extension<Arc<dyn Dataset>>,
+    axum::extract::Path(params): axum::extract::Path<HashMap<String, String>>,
+) -> Result<String, axum::http::StatusCode> {
+    let ref_param = params.get(PARAMETER_REF).unwrap();
+
+    let block_ref = match BlockRef::from_str(ref_param.as_str()) {
+        Ok(block_ref) => Ok(block_ref),
+        Err(_) => Err(axum::http::StatusCode::NOT_FOUND),
+    }?;
+
+    let get_ref_result = dataset.as_metadata_chain().get_ref(&block_ref).await;
+
+    match get_ref_result {
+        Ok(hash) => Ok(hash.to_string()),
+        Err(GetRefError::NotFound(_)) => Err(axum::http::StatusCode::NOT_FOUND),
+        Err(_) => {
+            tracing::debug!("Internal error while resolving reference '{}'", ref_param);
+            return Err(axum::http::StatusCode::INTERNAL_SERVER_ERROR);
+        }
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////
+
+pub async fn dataset_blocks_handler(
+    dataset: Extension<Arc<dyn Dataset>>,
+    axum::extract::Path(params): axum::extract::Path<HashMap<String, String>>,
```

Comment:
See my PR where I extract dataset id/name using a serde-derived struct - this should be much more efficient and cleaner than to access params via hashmap.

Reply:
Used. It really works with the named fields in structs, unlike tuple syntax, which expects all prefixes to be listed.